### PR TITLE
Limiting random MySQL server ids to 2^31-1.

### PR DESCRIPTION
**Pull request by alainjobart**
To fix https://github.com/youtube/vitess/issues/2280

### CODE REVIEW

File: go/vt/mysqlctl/mycnf_gen.go [hunk 1]
```diff
@@ -131,19 +131,27 @@ func (cnf *Mycnf) fillMycnfTemplate(tmplSrc string) (string, error) {
...
+// - It also avoids 1-99 because low numbers are used for fake slave
+// connections.  See NewSlaveConnection() in slave_connection.go for
+// more on that.
+// - It avoids the 2^31 - 2^32-1 range, as there seems to be some confusion
+// between the documentation at:
+// http://dev.mysql.com/doc/refman/5.7/en/replication-options.html
+// that implies is a full 32 bits number, and the code that encodes or
+// decodes serverIDs from core mysqld, which uses signed int32.
+// Such an ID may also be responsible for a mysqld crash in semi-sync code,
+// although we haven't been able to verify that yet.
```
Comment (michael-berlin) [inline go/vt/mysqlctl/mycnf_gen.go]: Can you please add the link to the GitHub issue here?

File: go/vt/mysqlctl/mycnf_gen.go [hunk 2]
```diff
@@ -131,19 +131,27 @@ func (cnf *Mycnf) fillMycnfTemplate(tmplSrc string) (string, error) {
...
-// See NewSlaveConnection() in slave_connection.go for more on that.
+// The value assigned to ServerID will be in the range [100, 2^31):
+// - It avoids 0 because that's reserved for mysqlbinlog dumps.
+// - It also avoids 1-99 because low numbers are used for fake slave
+// connections.  See NewSlaveConnection() in slave_connection.go for
+// more on that.
+// - It avoids the 2^31 - 2^32-1 range, as there seems to be some confusion
+// between the documentation at:
+// http://dev.mysql.com/doc/refman/5.7/en/replication-options.html
+// that implies is a full 32 bits number, and the code that encodes or
```
Comment (michael-berlin) [inline go/vt/mysqlctl/mycnf_gen.go]: > "that implies is a full ..."

nit: A word before "is" is missing?

File: go/vt/mysqlctl/mycnf_gen.go [hunk 3]
```diff
@@ -131,19 +131,27 @@ func (cnf *Mycnf) fillMycnfTemplate(tmplSrc string) (string, error) {
...
+// decodes serverIDs from core mysqld, which uses signed int32.
```
Comment (enisoc) [inline go/vt/mysqlctl/mycnf_gen.go]: I haven't seen any evidence that code in core mysqld uses signed int32. The problem I observed was that the semi-sync plugin appears to use signed int32.

File: go/vt/mysqlctl/mycnf_gen.go [hunk 4]
```diff
@@ -131,19 +131,28 @@ func (cnf *Mycnf) fillMycnfTemplate(tmplSrc string) (string, error) {
...
+// The value assigned to ServerID will be in the range [100, 2^31):
+// - It avoids 0 because that's reserved for mysqlbinlog dumps.
+// - It also avoids 1-99 because low numbers are used for fake slave
+// connections.  See NewSlaveConnection() in slave_connection.go for
+// more on that.
+// - It avoids the 2^31 - 2^32-1 range, as there seems to be some confusion
+// between the documentation at:
+// http://dev.mysql.com/doc/refman/5.7/en/replication-options.html
+// that implies serverID is a full 32 bits number, and the code that encodes or
+// decodes serverIDs from core mysqld, which uses signed int32.
```
Comment (enisoc) [inline go/vt/mysqlctl/mycnf_gen.go]: I haven't seen any evidence that code in core mysqld uses signed int32. The problem I observed was that the semi-sync plugin appears to use signed int32.

**Pull request merged**